torch/distributed/optim/optimizer.py

```diff
@@ -6,17 +6,25 @@
 
 
 class _LocalOptimizer:
+    # Ideally we would only need to share a lock for instances of
+    # _LocalOptimizer that deal with the same parameters. We are
+    # making a simplifying assumption here that if there is more
+    # than one instancde of _LocalOptimizer per worker, they will
+    # be optimizing the same parameters (e.g. each data parallel
+    # trainer will create its own instance of _LocalOptimizer but
+    # they will all optimize the same parameters on each worker)
```

Comment:
this is a strong assumption. Is there any workaround to drop this assumption and support per parameter lock?

Reply:
Workarounds would be too complex and not useful for most use-cases.